internal/driver/driver.go

```diff
@@ -206,86 +223,103 @@ outloop:
 	for i, req := range reqs {
+
 		s.lc.Debugf("S7Driver.HandleWriteCommands: protocols: %v, resource: %v, parameters: %v, attributes: %v", protocols, req.DeviceResourceName, params[i], req.Attributes)
 
-		variable := cast.ToString(req.Attributes["NodeName"])
-		dbInfo, _ := s.getDBInfo(variable)
+		var nodename = cast.ToString(req.Attributes["NodeName"])
+		var dbInfo, _ = s.getDBInfo(nodename)
 
 		reading, err := newCommandValue(req.Type, params[i])
 		if err != nil {
 			s.lc.Errorf("newCommandValue error: %s", err)
 		}
-		helper.SetValueAt(datas[i], 0, reading)
+		helper.SetValueAt(dataset[i], 0, reading)
 
 		// create gos7 DataItem
-		var item = gos7.S7DataItem{
+		var s7DataItem = gos7.S7DataItem{
 			Area:     dbInfo.Area,
 			WordLen:  dbInfo.WordLength,
 			DBNumber: dbInfo.DBNumber,
 			Start:    dbInfo.Start,
 			Amount:   dbInfo.Amount,
-			Data:     datas[i],
-			Error:    errors[i],
+			Data:     dataset[i],
+			Error:    s7_errors[i],
 		}
-		items = append(items, item)
+		s7DataItems = append(s7DataItems, s7DataItem)
 
 	}
-	fmt.Println("S7DataItems: ", items)
+
+	s.lc.Debugf("Write to S7DataItems: %s", s7DataItems)
 
 	// send command requests
 	times := int(reqs_len / batch_size)
 	remains := reqs_len % batch_size
-	var reqs1 []sdkModel.CommandRequest
-	var items1 = []gos7.S7DataItem{}
-	s7Client := s.getS7Client(deviceName)
+	var tmp_reqs []sdkModel.CommandRequest
+	var tmp_s7DateItems = []gos7.S7DataItem{}
+
+	s7Client := s.getS7Client(deviceName, protocols)
 
 outloop:
 	for j := 0; j <= times; j++ {
-		items1 = items1[:0]
-		reqs1 = reqs1[:0] // clear array
+
+		tmp_s7DateItems = tmp_s7DateItems[:0] // clear array
+		tmp_reqs = tmp_reqs[:0]               // clear array
 		if j >= times {
 			if remains > 0 {
-				items1 = items[times*batch_size : times*batch_size+remains]
+				tmp_s7DateItems = s7DataItems[times*batch_size : times*batch_size+remains]
 			} else {
 				break outloop // load complete, exit loop
 			}
 		} else {
-			items1 = items[j*batch_size : j*batch_size+batch_size]
+			tmp_s7DateItems = s7DataItems[j*batch_size : j*batch_size+batch_size]
 		}
-		// fmt.Println("reqs1: ", reqs1, "items1: ", items1)
 
-		// write data to S7
-		// fmt.Println("reqs len: ", len(items))
-		err = s7Client.Client.AGWriteMulti(items1, len(items1))
-		if err != nil {
-			s.lc.Errorf("AGWriteMulti Error: %s, reconnecting...", err)
-			s.mu.Lock()
-			s.s7Clients[deviceName] = s.NewS7Client(deviceName, protocols)
-			s.mu.Unlock()
-			return err
-		} else {
-			//fmt.Println(datas)
+		// write data to S7 device, if error, try 3 times
+		retrytimes := 3
+		for {
+			err = s7Client.Client.AGWriteMulti(tmp_s7DateItems, len(tmp_s7DateItems))
+			if err != nil {
+				s.lc.Errorf("AGWriteMulti Error: %s, reconnecting...", err)
+				s.mu.Lock()
+				s.s7Clients[deviceName] = nil
+				s.s7Clients[deviceName] = s.NewS7Client(deviceName, protocols)
+				s7Client = s.s7Clients[deviceName]
+				s.mu.Unlock()
```

Comment:
Use `getS7Client` since it handles nil and updates the map
```suggestion
				s.mu.Lock()
				s.s7Clients[deviceName] = nil
				s.mu.Unlock()
				ss7Client = s.getS7Client(deviceName, protocols)
				
```

Reply:
No, we can't get S7Client, because the tcp connection may be unreachable. Should be reconnect use NewS7Client function.

Reply:
getS7Client now calls NewS7Client when it is nil. See other comment.